File: backend/app/schemas/analytics.py
from pydantic import BaseModel


class DashboardStats(BaseModel):
    total_submissions: int
    total_judges: int
    total_reviews: int
    completed_reviews: int
    pending_reviews: int
    average_score: float | None


class SubmissionDistributionPoint(BaseModel):
    problem_statement: str
    count: int


class JudgeProgressPoint(BaseModel):
    judge_id: int
    judge_name: str
    completed: int
    pending: int
    total: int


class ScoreDistributionBucket(BaseModel):
    range_label: str
    count: int


class CriterionAveragePoint(BaseModel):
    criterion_id: int
    criterion_name: str
    average_score: float


class LeaderboardEntry(BaseModel):
    rank: int
    submission_id: int
    project_title: str
    problem_statement: str | None
    overall_score: float | None
    criterion_scores: dict[str, float]
    std_dev: float | None
    reviews_completed: int
    is_flagged: bool


class AnalyticsOverview(BaseModel):
    stats: DashboardStats
    submission_distribution: list[SubmissionDistributionPoint]
    judge_progress: list[JudgeProgressPoint]
    score_distribution: list[ScoreDistributionBucket]
    criterion_averages: list[CriterionAveragePoint]
    leaderboard_preview: list[LeaderboardEntry]

File: frontend/src/types/analytics.ts
export interface DashboardStats {
  total_submissions: number;
  total_judges: number;
  total_reviews: number;
  completed_reviews: number;
  pending_reviews: number;
  average_score: number | null;
}

/**
 * Review coverage. With no allocation step, coverage is emergent rather than guaranteed —
 * this is what tells an organiser whether every submission will actually get scored.
 */
export interface CoverageBucket {
  label: string;
  count: number;
}

export interface CoverageSubmission {
  id: string;
  project_title: string;
  team_identifier: string;
  /** Excludes reviews invalidated by a later edit — they provide no coverage. */
  completed_reviews: number;
  /** At least one completed review was invalidated when the team edited the submission. */
  needs_reevaluation: boolean;
  needed: number;
  is_scored: boolean;
}

export interface CoverageOut {
  scoring_limit: number;
  total_submissions: number;
  fully_scored: number;
  unreviewed: number;
  buckets: CoverageBucket[];
  submissions: CoverageSubmission[];
}

export interface JudgeProgressPoint {
  judge_id: number;
  judge_name: string;
  completed: number;
  pending: number;
  total: number;
}

export interface ScoreDistributionBucket {
  range_label: string;
  count: number;
}

export interface CriterionAveragePoint {
  criterion_id: number;
  criterion_name: string;
  average_score: number;
}

export interface LeaderboardEntry {
  rank: number;
  submission_id: number;
  project_title: string;
  overall_score: number | null;
  criterion_scores: Record<string, number>;
  std_dev: number | null;
  reviews_completed: number;
  is_flagged: boolean;
  /**
   * The team edited the submission after it was reviewed. `overall_score` and `std_dev` are
   * withheld (null) while this is true rather than reporting a figure computed from content
   * that has since changed; `reviews_completed` counts only the reviews that still apply.
   */
  needs_reevaluation: boolean;
}

export interface AnalyticsOverview {
  stats: DashboardStats;
  coverage: CoverageOut;
  judge_progress: JudgeProgressPoint[];
  score_distribution: ScoreDistributionBucket[];
  criterion_averages: CriterionAveragePoint[];
  leaderboard_preview: LeaderboardEntry[];
}
